src/tools/r2rdump/SignatureType.cs

```diff
@@ -38,20 +43,26 @@ public SignatureType(ref BlobReader signatureReader, MetadataReader mdReader, Ge
         {
             SignatureTypeCode signatureTypeCode = signatureReader.ReadSignatureTypeCode();
             Flags = 0;
-            if (signatureTypeCode == SignatureTypeCode.SZArray)
+
+            if (signatureTypeCode == SignatureTypeCode.ByReference)
             {
-                Flags |= SignatureTypeFlags.ARRAY;
+                Flags |= SignatureTypeFlags.REFERENCE;
                 signatureTypeCode = signatureReader.ReadSignatureTypeCode();
             }
-
-            TypeName = signatureTypeCode.ToString();
-            if (signatureTypeCode == SignatureTypeCode.TypeHandle || signatureTypeCode == SignatureTypeCode.ByReference)
+            if (signatureTypeCode == SignatureTypeCode.SZArray)
             {
-                if (signatureTypeCode == SignatureTypeCode.ByReference)
+                Flags |= SignatureTypeFlags.ARRAY;
+                ArrayDimension = 0;
+                while (signatureTypeCode == SignatureTypeCode.SZArray)
```

Comment:
This is checking for jagged array. It is not usual to deal with jagged arrays this way - the dimension of jagged is typically 1.

You may want to test it with a method signature like `static void method(int[][,,][,,,][][] x)` - verify that the signature is printed correctly.